index.rst

```diff
@@ -1,59 +1,367 @@
+
+ - ``/datasets/testing`` holds git LFS repositories that are used in CI and rarely change.  These are provided for convenience, and should be updated when their git master branches are; there will be no attempt to make old versions available.
+
+Finally, we propose that all raw ingestion into the shared repository be done with symbolic links or in-place (outside-root) ingests, pointing to read-only filesystems - the existing ``/lsstdata`` for Rubin Observatory data, and a new ``/external-raw`` filesystem for raw data from other instruments.
+
+.. note ::
+
+   **TODO**: This document should probably say more about staging locations for (at least) raws that haven't been ingested yet, for both Rubin Observatory instruments and precursor instruments (which I imagine might be quite different in this respect).  I'd like to at least be confident that the filesystem locations and permissions I'm proposing are consistent with how data arrives and gets ingested.  I don't know enough to write that up myself.
+
+.. _access-controls:
+
+Access Controls
+===============
+
+The current Gen3 registry architecture does not allow any fine-grained access control in the repository database; we instead rely on "friendly users" being careful and respectful of this shared space.
+
+At the same time, we will use filesystem access controls to protect shared and per-user files, and we plan to implement some checks in the Butler client itself to make it at least extremely difficult to *accidentally* cause problems.
+
+.. warning::
+
+   NEVER use ``psql`` or other direct-SQL clients (e.g. the Python DBAPI or SQLAlchemy) to perform write operations in the repository database.
+   These can corrupt the data repository, and we have essentially no way to guard against them.
+
+   It should not be necessary in the long term to ever use direct SQL access even for read access; the SQL schema is *not* considered a public interface - but we recognize that this may be necessary for debugging for a while.
+   This can be ensured by running::
+
+      SET SESSION CHARACTERISTICS AS TRANSACTION READ ONLY;
+
+   at the start of the session.
+
+   If you have to do this (and not at the prompting of a middleware team member trying to help diagnose a problem), please also create a ticket explaining what you wanted to do that couldn't be done with butler tools, so we can address that feature gap.
+
+This proposal specifies filesystem access controls in terms of a number of high-level "roles" that certain operators or developers may temporarily opt in to via ``su`` or special setuid tools.
+Usually these roles will be used only to create subdirectories that are owned directly by the true user.
```

Comment:
Does "true user" here mean "the user invoking the command"? Thinking about it, does it matter what e.g. the file ownership is, or what user/role runs the command? Does the registry track the user that created a dataset? If so, then we almost certainly do want the user, not the role, to run the pipeline and write the data. I don't know how that would work in actual setup.

Reply:
> Does "true user" here mean "the user invoking the command"?

Yes.

> Does the registry track the user that created a dataset?

No; some future Science Platform version might, but right now we don't track ownership of anything in the database itself.  I'm hoping to at most _simulate_ user ownership of collections via naming conventions ("if a collection starts with `u/jbosch`, it's owned by jbosch"), and for `RUN` collections that ownership would naturally extend to the datasets within.

> Thinking about it, does it matter what e.g. the file ownership is, or what user/role runs the command?

I think it matters only in terms of who is allowed to go back and delete them later, for which the answer is usually "anyone who can assume the role that created them" (unless it's "no one, ever"), but I think we can enable either with personal ownership or role ownership depending on how we handle groups, and that's another thing I'd like to leave to the sysadmins.